[Nullability Annotations to Java Classes] Replace `findViewById` with `ViewBinding` for History (`relatively safe`)

ParaskP7 opened the pull request:
Parent #18906

This PR's main focus is to replace `findViewById ` with `ViewBinding` for the [HistoryDetailContainerFragment.java](https://github.com/wordpress-mobile/WordPress-Android/blob/5b0d3f39c8cfe9c08b4e1547e46bc947f9be6b8f/WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java#L61) class.

Additionally, this PR is also dealing with adding any missing nullability annotations to this fragment class, focusing mainly on its fields, and, any effected neighbour classes. But first, all existing warnings on this class are being resolved/suppressed so as to make it easier to deal with any missing nullability checks (with help from the IDE), especially when a `@Nullable` annotation is added.

FYI: This change is `relatively safe`, meaning that although there are compile-time changes associated with this change, and it needs testing, the changes included in this PR are very targeted and specific to one screen, and one screen only. Thus, if this screen is tested appropriately, it should be safe to merge this to `trunk`.

-----

PS: @RenanLukas I added you as the main reviewer, randomly so, since I just wanted someone from the Jetpack/WordPress mobile team to be aware of and sign-off on that change for JP/WPAndroid. Feel free to merge this PR directly yourself if you deem so.

-----

`findViewById ` -> `ViewBinding`

1. [Replace findviewbyid with viewbinding for history detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19172/commits/85f6586cd0d079ec28b4ab0564fbada8f9d4affd)
2. [Use non-null viewbinding method parameter for history detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19172/commits/6512ed28f0ab2466de6f3a824258198a98fa680a)

Nullability Annotation List:

1. [Add missing nullability annotations to sdk override methods](https://github.com/wordpress-mobile/WordPress-Android/pull/19172/commits/7c204435590d6aa2250811383f37337d0ba46209)
2. [Remove adapter from field and use it as non-null](https://github.com/wordpress-mobile/WordPress-Android/pull/19172/commits/e28be1d240e52e437756938bbcbcb7709e198c62)
3. [Remove revisions from field and use it as nullable](https://github.com/wordpress-mobile/WordPress-Android/pull/19172/commits/01bb24e9d2976ce67242b397cffaf1c1ec9f2ebc)
4. [Add missing nl-a to on page change listener field](https://github.com/wordpress-mobile/WordPress-Android/pull/19172/commits/e8112617b0b752d423f9e9e5c2ad0f5ecdf10d26)
5. [Add missing nl-a to on revision field](https://github.com/wordpress-mobile/WordPress-Android/pull/19172/commits/102dc63f4b2cec66e7470ee2f5643e8af3e57a78)

Warnings Resolution List:

1. [Make inner history detail fragment adapter class static](https://github.com/wordpress-mobile/WordPress-Android/pull/19172/commits/16d535f411db63703cf30a0a9d4b1b666f8df353)
2. [Replace anonymous class with lambda for history detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19172/commits/5e643b5347cd6f65a98ee3da3515837a8197a5f5)

Warnings Suppression List:

1. [Suppress deprecation warnings on history detail](https://github.com/wordpress-mobile/WordPress-Android/pull/19172/commits/c508f268f75be002b5a772f55088f2a2dac80aaf)

-----

## To test:

1. Quickly smoke test any history related functionality on both, the WordPress and Jetpack apps, and see if everything is working as expected.
8. In addition to the above smoke test, you can expand the below and follow the inner and more explicit test steps within:

<details>
    <summary>History Details Screen [HistoryDetailContainerFragment.java]</summary>

ℹ️ This test applies to both, the `Jetpack` and `WordPress` app.

- Go to `Posts` screen and tap on any post.
- Find the options menu (`top-right`) and tap on `History` .
- From the `History` screen, tap on any revision.
- Verify that the `History Details` screen is shown and functioning as expected.

</details>

-----

## Regression Notes

1. Potential unintended areas of impact

    - N/A

6. What I did to test those areas of impact (or what existing automated tests I relied on)

    - See `To test` section above.

9. What automated tests I added (or what prevented me from doing so)

    - N/A

-----

## PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

ParaskP7 committed:
Refactor: Replace findviewbyid with viewbinding for history detail

ParaskP7 committed:
Refactor: Use non-null viewbinding method parameter for history detail

This is done in order to avoid using the 'if (mBinding != null) { ... }'
check everywhere and only use that where absolutely necessary, which
usually is only needed when a 'binding' is required via an SDK related
'@Override' method, that is, it being the 'binding' starting point.

FYI: This way, when this class gets converted to Kotlin (in the future),
instead of using the 'binding' as a parameter to a method, the same
method will get converted to a 'binding' extension function, and then,
everything will become much more straightforward. For what is worth,
this pattern is already used as such with Kotlin classes. For example,
see 'PreviewImageFragment.kt' class and its 'initializeViews()'
extension function.

ParaskP7 committed:
Analysis: Add missing nullability annotations to sdk override methods

ParaskP7 committed:
Analysis: Suppress deprecation warnings on history detail

Warnings:
- "'setHasOptionsMenu(boolean)' is deprecated"
- "Overrides deprecated method in 'androidx.fragment.app.Fragment'"
- "'androidx.fragment.app.FragmentStatePagerAdapter' is deprecated"
- "'FragmentStatePagerAdapter(androidx.fragment.app.FragmentManager)'
is deprecated "

ParaskP7 committed:
Analysis: Make inner history detail fragment adapter class static

Warning: "Inner class 'HistoryDetailFragmentAdapter' may be 'static'"

ParaskP7 committed:
Analysis: Replace anonymous class with lambda for history detail

Warning: "Anonymous new Runnable() can be replaced with lambda"

ParaskP7 committed:
Analysis: Remove adapter from field and use it as non-null

ParaskP7 committed:
Analysis: Remove revisions from field and use it as nullable

ParaskP7 committed:
Analysis: Add missing nl-a to on page change listener field

FYI: 'nl-a' stands for 'nullable annotation'.

ParaskP7 committed:
Analysis: Add missing nl-a to on revision field

FYI: 'nl-a' stands for 'nullable annotation'.

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java:
```diff
@@ -100,85 +88,83 @@ public static HistoryDetailContainerFragment newInstance(final Revision revision
-        resetOnPageChangeListener();
-
-        return rootView;
+        return mBinding.getRoot();
```
💡 We're checking if `mBinding` is `null` on line `115` but we're using it here without checking it. I wonder if we should do something here if `mBinding` is `null` (e.g. throwing an exception) or if we can just drop the null check in line `115` 🤔

ParaskP7 replied:
Yeaaa, and thanks for this comment @RenanLukas , ideally I would like to drop the null check on line `115`. Unfortunately, when I do that, for some reason, the IDE is warning us about it, I just doesn't understand that `mBinding` is already set, just a few lines above: `Dereference of 'mBinding' may produce 'NullPointerException'`

Also, when I do that, that is, dropping the null check on line `115`, the IDE now starts warning us about line `163`, as expected, I don't know why it is warning us on that already, probably an IDE bug of some kind: `Method invocation 'getRoot' may produce 'NullPointerException' `

🤷 🤷 🤷

I ended-up choosing to throw an `IllegalStateException` and guard the `mBinding.getRoot()` call. Let me know what you think and again, thanks for this suggestion:

This is done here: 724f91b43d1342b4fc4a50da144359a202b9ca58

RenanLukas reviewed (commented):
Thanks for the PR, @ParaskP7. I've tested WP and JP and everything is working as expected 🎉 

I've left a small comment about the `mBinding` null check. Please let me know if it makes sense.

ParaskP7 committed:
Analysis: Throw illegal state exception if binding or revision is null

This is done in order to include the 'mBinding.getRoot()' return within
the 'mBinding' nullability check and avoid getting a NPE on it, just
like it is already done for every 'mBinding' related call above.

PR Comment: https://github.com/wordpress-mobile/WordPress-Android/
pull/19172#discussion_r1329379578

ParaskP7 commented:
Thanks for reviewing and testing this @RenanLukas , much appreciated! 🙇 ❤️ 🚀

RenanLukas reviewed (approved):
Thanks for updating the PR, @ParaskP7. LGTM :shipit: 

Considering the PR instructions I'm going to merge it to `trunk`.

ParaskP7 commented:
Thank you for merging @RenanLukas ! 🙇 ❤️ 🚀
